datadog/resource_datadog_service_level_objective.go

```diff
@@ -151,12 +151,110 @@ func ValidateServiceLevelObjectiveTypeString(v interface{}, k string) (ws []stri
 
 func buildServiceLevelObjectiveStruct(d *schema.ResourceData) *datadogV1.ServiceLevelObjective {
 
-	slo := &datadogV1.ServiceLevelObjective{
-		Id:   datadogV1.PtrString(d.Id()),
-		Name: d.Get("name").(string),
-		Type: datadogV1.SLOType(d.Get("type").(string)),
+	slo := datadogV1.NewServiceLevelObjectiveWithDefaults()
+	slo.SetName(d.Get("name").(string))
+	slo.SetType(datadogV1.SLOType(d.Get("type").(string)))
+	slo.SetId(d.Id())
+
+	if attr, ok := d.GetOk("description"); ok {
+		slo.SetDescription(attr.(string))
+	}
+
+	switch slo.GetType() {
+	case datadogV1.SLOTYPE_MONITOR:
+		// monitor type
+		if attr, ok := d.GetOk("monitor_ids"); ok {
+			s := make([]int64, 0)
+			for _, v := range attr.(*schema.Set).List() {
+				s = append(s, int64(v.(int)))
+			}
+			slo.SetMonitorIds(s)
+		}
+		if attr, ok := d.GetOk("groups"); ok {
+			s := make([]string, 0)
+			for _, v := range attr.(*schema.Set).List() {
+				s = append(s, v.(string))
+			}
+			slo.SetGroups(s)
+		}
+	default:
+		// metric type
+		if attr, ok := d.GetOk("query"); ok {
+			queries := make([]map[string]interface{}, 0)
+			raw := attr.([]interface{})
+			for _, rawQuery := range raw {
+				if query, ok := rawQuery.(map[string]interface{}); ok {
+					queries = append(queries, query)
+				}
+			}
+			if len(queries) >= 1 {
+				// only use the first defined query
+				slo.SetQuery(*datadogV1.NewServiceLevelObjectiveQuery(
+					queries[0]["denominator"].(string),
+					queries[0]["numerator"].(string)))
+			}
+		}
+	}
+
+	if attr, ok := d.GetOk("tags"); ok {
+		s := make([]string, 0)
+		for _, v := range attr.(*schema.Set).List() {
+			s = append(s, v.(string))
+		}
+		slo.SetTags(s)
 	}
 
+	if _, ok := d.GetOk("thresholds"); ok {
+		numThresholds := d.Get("thresholds.#").(int)
+		sloThresholds := make([]datadogV1.SLOThreshold, 0)
+		for i := 0; i < numThresholds; i++ {
+			prefix := fmt.Sprintf("thresholds.%d.", i)
+			t := datadogV1.NewSLOThresholdWithDefaults()
+
+			if tf, ok := d.GetOk(prefix + "timeframe"); ok {
+				t.SetTimeframe(datadogV1.SLOTimeframe(tf.(string)))
+			}
+
+			if targetValue, ok := d.GetOk(prefix + "target"); ok {
+				if f, ok := floatOk(targetValue); ok {
+					t.SetTarget(f)
+				}
+			}
+
+			if warningValue, ok := d.GetOk(prefix + "warning"); ok {
+				if f, ok := floatOk(warningValue); ok {
+					t.SetWarning(f)
+				}
+			}
+
+			if targetDisplayValue, ok := d.GetOk(prefix + "target_display"); ok {
+				if s, ok := targetDisplayValue.(string); ok && strings.TrimSpace(s) != "" {
+					t.SetTargetDisplay(strings.TrimSpace(targetDisplayValue.(string)))
+				}
+			}
+
+			if warningDisplayValue, ok := d.GetOk(prefix + "warning_display"); ok {
+				if s, ok := warningDisplayValue.(string); ok && strings.TrimSpace(s) != "" {
+					t.SetWarningDisplay(strings.TrimSpace(warningDisplayValue.(string)))
```

Comment:
Can be later/not as part of this PR, but I wonder if a StateFunc or ValidateFunc on these to trim spaces would help here - terraform.io/docs/extend/schemas/schema-behaviors.html#statefunc